### PR TITLE
{Q,Z},of_string accept underscores

### LINKED ISSUE #71: Make Z.of_string and Q.of_string support underscores
Request extracted from #10

### PR DESCRIPTION
**Pull request by hhugo**
close #71

### CODE REVIEW

**Comment by hhugo** [comment]
rebased on top of the current master branch

**Comment by hhugo** [comment]
I've done some cleanup and added more tests.

File: caml_z.c [hunk 1]
```diff
@@ -637,22 +639,23 @@ CAMLprim value ml_z_of_substring_base(value b, value v, value offset, value leng
...
-    /* make sure that dd is nul terminated */
-    dd[sz] = 0;
-    for (i = 0; i < sz; i++) {
-      if (dd[i] >= '0' && dd[i] <= '9') dd[i] -= '0';
-      else if (dd[i] >= 'a' && dd[i] <= 'f') dd[i] -= 'a' - 10;
-      else if (dd[i] >= 'A' && dd[i] <= 'F') dd[i] -= 'A' - 10;
+    char* digits = (char*)malloc(sz+1);
+    for (i = 0, j = 0; i < sz; i++, j++) {
+      /* skip underscores but leading ones */
+      if (i > 0 && d[i] == '_') {j--;continue;};
```
Comment (xavierleroy) [inline caml_z.c:645]: It's a matter of taste, but I would rather not increment `j` in the "next" part of the `for`, rather at the end of the body:
```
    for (i = 0, j = 0; i < sz; i++) {
      /* skip underscores but leading ones */
      if (i > 0 && d[i] == '_') continue;
      ....
      j++;
    }
```

**Review by xavierleroy — COMMENTED** [review]
Looks good to me.  Thanks for the tests!  A minor style suggestion below.

**Comment by antoinemine** [comment]
Hi !
I agree with Xavier's remark (I'm especially not fond of the `j--` before the `continue`).
Also, we should free digits in case of `caml_invalid_argument` to avoid memory leaks.

**Comment by hhugo** [comment]
- I've applied your suggestion
- freed digits caml_invalid_arguement
- cleaned up the implementation further

**Comment by hhugo** [comment]
hold on merging this one, I'm investigating some potential bug.

**Comment by hhugo** [comment]
False alert.  ready to merge.

**Review by xavierleroy — APPROVED** [review]
Looks good to me, thanks.

**Pull request merged**